File: backend/cmd/orva/adapters/node24/orva.js
// Orva Node.js SDK — kv, invoke, jobs.
//
// Routes through ORVA_API_BASE (loopback) using ORVA_INTERNAL_TOKEN that
// was injected at worker spawn. Both env vars must be present in
// production; absent in tests where the SDK throws OrvaUnavailableError.
//
// Usage:
//   const { kv, invoke, jobs } = require('orva')
//   await kv.put('count', 42, { ttlSeconds: 60 })
//   const n = await kv.get('count', 0)
//   const result = await invoke('resize-image', { url: '...' })
//   await jobs.enqueue('send-welcome-email', { to: 'ada@example.com' })

class OrvaError extends Error {
  constructor(message, status = 0) {
    super(message)
    this.name = 'OrvaError'
    this.status = status
  }
}

class OrvaUnavailableError extends OrvaError {
  constructor(message) {
    super(message, 0)
    this.name = 'OrvaUnavailableError'
  }
}

const _apiBase = () => process.env.ORVA_API_BASE || ''
const _token   = () => process.env.ORVA_INTERNAL_TOKEN || ''
const _fnID    = () => process.env.ORVA_FUNCTION_ID || ''

async function _request(method, path, { body = null, headers = {} } = {}) {
  const base  = _apiBase()
  const token = _token()
  if (!base || !token) {
    throw new OrvaUnavailableError(
      'Orva SDK not available (missing ORVA_API_BASE or ORVA_INTERNAL_TOKEN)'
    )
  }
  const init = {
    method,
    headers: {
      'X-Orva-Internal-Token': token,
      'Content-Type': 'application/json',
      ...headers,
    },
  }
  if (body != null) {
    init.body = typeof body === 'string' ? body : JSON.stringify(body)
  }
  const res  = await fetch(base + path, init)
  const text = await res.text()
  return { status: res.status, body: text }
}

// ── KV ──────────────────────────────────────────────────────────────

const kv = {
  /**
   * Read a value previously stored with kv.put. Returns the default
   * when the key is missing or expired. Values are JSON-decoded.
   */
  async get(key, defaultValue = null) {
    const fn = _fnID()
    const { status, body } = await _request('GET', `/api/v1/_kv/${fn}/${key}`)
    if (status === 404) return defaultValue
    if (status >= 400)  throw new OrvaError(`kv.get(${key}) failed: ${body}`, status)
    const data = JSON.parse(body)
    return data.value != null ? JSON.parse(data.value) : defaultValue
  },

  /**
   * Store a value. value is JSON-encoded. opts.ttlSeconds=0 (default)
   * means no expiry; positive means expire after that many seconds.
   */
  async put(key, value, { ttlSeconds = 0 } = {}) {
    const fn = _fnID()
    const payload = { value: JSON.stringify(value), ttl_seconds: ttlSeconds | 0 }
    const { status, body } = await _request('PUT', `/api/v1/_kv/${fn}/${key}`, { body: payload })
    if (status >= 400) throw new OrvaError(`kv.put(${key}) failed: ${body}`, status)
  },

  async delete(key) {
    const fn = _fnID()
    const { status, body } = await _request('DELETE', `/api/v1/_kv/${fn}/${key}`)
    if (status >= 400 && status !== 404) {
      throw new OrvaError(`kv.delete(${key}) failed: ${body}`, status)
    }
  },

  async list({ prefix = '', limit = 100 } = {}) {
    const fn  = _fnID()
    const qs  = new URLSearchParams()
    qs.set('limit', String(limit))
    if (prefix) qs.set('prefix', prefix)
    const { status, body } = await _request('GET', `/api/v1/_kv/${fn}?${qs}`)
    if (status >= 400) throw new OrvaError(`kv.list failed: ${body}`, status)
    const data = JSON.parse(body)
    // value is already a JSON object on the wire; no second parse needed.
    return data.keys || []
  },
}

// ── Function-to-function invoke ─────────────────────────────────────

/**
 * Invoke another Orva function by friendly name. Returns the parsed
 * {statusCode, headers, body} envelope. body is JSON-decoded when
 * possible; remains a string otherwise.
 */
async function invoke(functionName, payload = {}) {
  const headers = {}
  // Forward inbound call depth so nested invokes hit the cap.
  const incoming = process.env.ORVA_CALL_DEPTH
  if (incoming) headers['X-Orva-Call-Depth'] = incoming

  const { status, body } = await _request(
    'POST',
    `/api/v1/_internal/invoke/${functionName}`,
    { body: payload, headers }
  )

  if (status === 404) throw new OrvaError(`function not found: ${functionName}`, 404)
  if (status === 507) throw new OrvaError('call depth exceeded', 507)
  if (status >= 400)  throw new OrvaError(`invoke(${functionName}) failed: ${body}`, status)

  const env = JSON.parse(body)
  if (typeof env.body === 'string') {
    try {
      env.body = JSON.parse(env.body)
    } catch {
      // leave as string
    }
  }
  return env
}

// ── Background jobs ─────────────────────────────────────────────────

const jobs = {
  /** Enqueue a fire-and-forget job. Returns the job id. */
  async enqueue(functionName, payload = {}, { maxAttempts = 3, scheduledAt = null } = {}) {
    const bodyObj = {
      function_name: functionName,
      payload,
      max_attempts: maxAttempts | 0,
    }
    if (scheduledAt) bodyObj.scheduled_at = scheduledAt

    const { status, body } = await _request('POST', '/api/v1/jobs', { body: bodyObj })
    if (status >= 400) throw new OrvaError(`jobs.enqueue failed: ${body}`, status)
    return JSON.parse(body).id
  },
}

module.exports = { kv, invoke, jobs, OrvaError, OrvaUnavailableError }

File: backend/runtimes/python313/orva.py
"""Orva Python SDK — kv, invoke, jobs.

Available inside any function running on Orva. Routes through the
ORVA_API_BASE loopback URL using the per-process ORVA_INTERNAL_TOKEN
that the worker received at spawn time. Both env vars are present in
production and absent in tests; helpers raise OrvaUnavailableError when
the SDK can't reach the host.

Usage:

    from orva import kv, invoke, jobs

    kv.put("count", 42, ttl_seconds=60)
    n = kv.get("count", default=0)

    result = invoke("resize-image", {"url": "..."})

    jobs.enqueue("send-welcome-email", {"to": "ada@example.com"})
"""

from __future__ import annotations

import json
import os
import urllib.error
import urllib.request


class OrvaError(RuntimeError):
    """Base class for SDK errors. Status codes from the upstream HTTP
    response surface as `status` so callers can branch."""

    def __init__(self, message: str, status: int = 0):
        super().__init__(message)
        self.status = status


class OrvaUnavailableError(OrvaError):
    """ORVA_API_BASE / ORVA_INTERNAL_TOKEN are missing — the SDK can't
    reach the host. Typically only happens in tests or when running the
    handler outside of Orva's sandbox."""


def _api_base() -> str:
    return os.environ.get("ORVA_API_BASE", "")


def _token() -> str:
    return os.environ.get("ORVA_INTERNAL_TOKEN", "")


def _function_id() -> str:
    return os.environ.get("ORVA_FUNCTION_ID", "")


def _trace_headers() -> dict:
    """Forward trace context on every internal call so F2F / job enqueues
    stay linked into the same trace as the caller. Empty when the
    function wasn't started inside a trace (legacy or test)."""
    h: dict = {}
    if v := os.environ.get("ORVA_TRACE_ID"):
        h["X-Orva-Trace-Id"] = v
    if v := os.environ.get("ORVA_SPAN_ID"):
        h["X-Orva-Span-Id"] = v
    if v := _function_id():
        h["X-Orva-Caller-Function"] = v
    return h


def _request(method: str, path: str, *, body: bytes = b"", headers: dict | None = None) -> tuple[int, bytes]:
    base = _api_base()
    token = _token()
    if not base or not token:
        raise OrvaUnavailableError("Orva SDK not available (missing ORVA_API_BASE or ORVA_INTERNAL_TOKEN)")

    h = {"X-Orva-Internal-Token": token, "Content-Type": "application/json"}
    h.update(_trace_headers())
    if headers:
        h.update(headers)

    req = urllib.request.Request(base + path, data=body or None, headers=h, method=method)
    try:
        with urllib.request.urlopen(req, timeout=30) as resp:
            return resp.getcode(), resp.read()
    except urllib.error.HTTPError as e:
        return e.code, e.read()


# ── KV ──────────────────────────────────────────────────────────────


class _KV:
    """Per-function key/value store backed by SQLite on the host. Values
    are JSON-encoded transparently. TTL is in seconds; 0 (default) means
    the key persists until deleted."""

    @staticmethod
    def get(key: str, default=None):
        fn = _function_id()
        status, body = _request("GET", f"/api/v1/_kv/{fn}/{key}")
        if status == 404:
            return default
        if status >= 400:
            raise OrvaError(f"kv.get({key!r}) failed: {body!r}", status=status)
        data = json.loads(body)
        return json.loads(data["value"]) if data.get("value") is not None else default

    @staticmethod
    def put(key: str, value, ttl_seconds: int = 0) -> None:
        fn = _function_id()
        payload = json.dumps({
            "value": json.dumps(value),
            "ttl_seconds": int(ttl_seconds),
        }).encode("utf-8")
        status, body = _request("PUT", f"/api/v1/_kv/{fn}/{key}", body=payload)
        if status >= 400:
            raise OrvaError(f"kv.put({key!r}) failed: {body!r}", status=status)

    @staticmethod
    def delete(key: str) -> None:
        fn = _function_id()
        status, body = _request("DELETE", f"/api/v1/_kv/{fn}/{key}")
        if status >= 400 and status != 404:
            raise OrvaError(f"kv.delete({key!r}) failed: {body!r}", status=status)

    @staticmethod
    def list(prefix: str = "", limit: int = 100):
        fn = _function_id()
        path = f"/api/v1/_kv/{fn}?limit={int(limit)}"
        if prefix:
            path += f"&prefix={prefix}"
        status, body = _request("GET", path)
        if status >= 400:
            raise OrvaError(f"kv.list failed: {body!r}", status=status)
        data = json.loads(body)
        # The wire format already returns each entry's value as the
        # original JSON object — no second decode needed.
        return list(data.get("keys") or [])


kv = _KV()


# ── Function-to-function invoke ─────────────────────────────────────


def invoke(function_name: str, payload=None, *, timeout: int = 30):
    """Invoke another Orva function by friendly name. Returns the parsed
    {statusCode, headers, body} envelope so callers can branch on
    statusCode if they care. The body is JSON-decoded when possible."""

    body = json.dumps(payload if payload is not None else {}).encode("utf-8")
    headers = {}
    # Forward call depth so nested invokes cap out before deadlocking the pool.
    incoming = os.environ.get("ORVA_CALL_DEPTH", "")
    if incoming:
        headers["X-Orva-Call-Depth"] = incoming
    status, raw = _request("POST", f"/api/v1/_internal/invoke/{function_name}", body=body, headers=headers)
    if status == 404:
        raise OrvaError(f"function not found: {function_name}", status=404)
    if status == 507:
        raise OrvaError("call depth exceeded", status=507)
    if status >= 400:
        raise OrvaError(f"invoke({function_name!r}) failed: {raw!r}", status=status)

    envelope = json.loads(raw)
    body_str = envelope.get("body", "")
    if isinstance(body_str, str):
        try:
            envelope["body"] = json.loads(body_str)
        except (ValueError, TypeError):
            pass  # leave as string
    return envelope


# ── Background jobs ─────────────────────────────────────────────────


class _Jobs:
    """Fire-and-forget queue. enqueue() returns the job id."""

    @staticmethod
    def enqueue(function_name: str, payload=None, *, max_attempts: int = 3, scheduled_at=None) -> str:
        body_obj = {
            "function_name": function_name,
            "payload": payload if payload is not None else {},
            "max_attempts": int(max_attempts),
        }
        if scheduled_at is not None:
            body_obj["scheduled_at"] = scheduled_at
        status, raw = _request("POST", "/api/v1/jobs", body=json.dumps(body_obj).encode("utf-8"))
        if status >= 400:
            raise OrvaError(f"jobs.enqueue failed: {raw!r}", status=status)
        data = json.loads(raw)
        return data["id"]


jobs = _Jobs()


__all__ = ["kv", "invoke", "jobs", "OrvaError", "OrvaUnavailableError"]
